### DISCUSSION TITLE
Welcome to pico-cec Discussions!

### DISCUSSION THREAD

**Discussion by gkoh** [gkoh/pico-cec#12] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

**Comment by LaserBearIndustries** [comment]
Thank you for this amazing project! I was working on something similar for a while now to pair with the MiSTer FPGA gaming platform and this solved a lot of the problems I was facing getting reliability from CEC connections. 

I am working with a custom PCB and the RP2040 Zero from Waveshare, mainly for the larger buttons to make a shell with integrated buttons work a bit easier. 
![CEC_Enabler](https://github.com/user-attachments/assets/986e12eb-6785-4870-b2f6-a87930c26ec9)
I am going to sell the hardware and have made some changes to the code for use with the MiSTer FPGA gaming platform since the DE-10 Nano board has a hardware problem where it crashes all CEC activity on the TV. The custom PCB just does not connect the CEC pin to the device side which has solved the main problem. 

I have started on code to use the WS2812 LED on the RP2040 Zero. I have it functioning for the blink task, and the to indicate when a button is pressed. I also added support to change the blink task output as I want to make it go green when it connects to CEC tv side, and to stop the blinking in the event of an error state. If you have any pointers on how to make that work I would love to hear them.

WS2812 is handled in PIO, I haven't made it portable based on the board yet as I am targeting my hardware.  

You can see those changes in my fork if you are interested at all, also if you have a donation link let me know or if you would like one of my dongles I would be happy to send it your way!

**Reply by gkoh** [reply]
@LaserBearIndustries

> Thank you for this amazing project! I was working on something similar for a while now to pair with the MiSTer FPGA gaming platform and this solved a lot of the problems I was facing getting reliability from CEC connections.

Thank you for the note!
Just recently I was motivated enough to run the Linux CEC compliance test and was somewhat impressed that it mostly passed.

> I am going to sell the hardware and have made some changes to the code for use with the MiSTer FPGA gaming platform since the DE-10 Nano board has a hardware problem where it crashes all CEC activity on the TV. The custom PCB just does not connect the CEC pin to the device side which has solved the main problem.
> 
> I have started on code to use the WS2812 LED on the RP2040 Zero. I have it functioning for the blink task, and the to indicate when a button is pressed. I also added support to change the blink task output as I want to make it go green when it connects to CEC tv side, and to stop the blinking in the event of an error state. If you have any pointers on how to make that work I would love to hear them.
> 
> WS2812 is handled in PIO, I haven't made it portable based on the board yet as I am targeting my hardware.

I spent a bit of time reviewing LaserBear's CEC-Enabler (both various online progress posts and the github fork).
I am truly honoured that the code is considered robust (and valuable) enough to be commercialised in any capacity.

The changes in the CEC-Enabler fork are minimal and non-invasive (if anything, my recent refactor was significant and disruptive to downstreams projects (something I had not considered)).

I propose the following as a way to ensure CEC-Enabler is viable and the community benefits:
* I will incorporate and improve the WS2812 changes you made, that feature has been on the to-do since inception
   * I'm 100% OK with PIO, looking at the timing requirements for WS2812 this is the fastest path to a robust solution on RP2040
      *  platform-compatibility has not been the highest priority (and can be handled if/when demand for another platform arises)
* I can expose any critical hardware dependent items in the build system (esp WS2812 pin) such that building for a different target (eg CEC-Enabler) is mostly trivial
* I can incorporate the `MiSTer` keymap into the `pico-cec` code base
   * I can add a `#define` build time control for choosing the default keymap
* All the above should mean CEC-Enabler is easier to maintain as a downstream project
   * I'd encourage any non-MiSTer target improvements be upstreamed to `pico-cec`
   * CEC-Enabler continues to be a separate code-base managed by LaserBear

> You can see those changes in my fork if you are interested at all, also if you have a donation link let me know or if you would like one of my dongles I would be happy to send it your way!

This is a very generous offer!
Please email me at `guorong.koh _at_ gmail.com` to discuss obtaining a dongle.

PS. The hardware looks nice and robust (much more so than my feeble design skills allow).
